Userland/Libraries/LibWeb/Infra/Strings.cpp

```diff
@@ -29,4 +31,38 @@ String strip_and_collapse_whitespace(StringView string)
     return builder.string_view().trim(Infra::ASCII_WHITESPACE);
 }
 
+// https://infra.spec.whatwg.org/#code-unit-prefix
+bool is_code_unit_prefix(StringView potential_prefix, StringView input)
+{
+    auto potential_prefix_utf16 = utf8_to_utf16(potential_prefix);
+    auto input_utf16 = utf8_to_utf16(input);
+
+    // 1. Let i be 0.
+    size_t i = 0;
+
+    // 2. While true:
+    while (true) {
+        // 1. If i is greater than or equal to potentialPrefix’s length, then return true.
+        if (i >= potential_prefix.length())
+            return true;
+
+        // 2. If i is greater than or equal to input’s length, then return false.
+        if (i >= input.length())
+            return false;
+
+        // 3. Let potentialPrefixCodeUnit be the ith code unit of potentialPrefix.
+        auto potential_prefix_code_unit = Utf16View(potential_prefix_utf16).code_unit_at(i);
+
+        // 4. Let inputCodeUnit be the ith code unit of input.
+        auto input_code_unit = Utf16View(input_utf16).code_unit_at(i);
+
+        // 5. Return false if potentialPrefixCodeUnit is not inputCodeUnit.
+        if (potential_prefix_code_unit != input_code_unit)
+            return false;
+
+        // 6. Set i to i + 1.
+        ++i;
+    }
```

Comment:
I understand we want to follow the spec but this code makes this seems much more complicated than the startsWith it is

Reply:
Also why do we need to convert to utf16?
The codepoints have to match exactly right?

I might be missing something obvious

Reply:
> When it is clear from context that [code units](https://infra.spec.whatwg.org/#code-unit) are in play, e.g., because one of the strings is a literal containing only characters that are in the range U+0020 SPACE to U+007E (~), "input starts with potentialPrefix" can be used as a synonym for "potentialPrefix is a [code unit prefix](https://infra.spec.whatwg.org/#code-unit-prefix) of input".

> With unknown values, it is good to be explicit: targetString is a [code unit prefix](https://infra.spec.whatwg.org/#code-unit-prefix) of userInput. But with a literal, we can use plainer language: userInput [starts with](https://infra.spec.whatwg.org/#string-starts-with) "!".

I assume this is supposed to remove any reliance on starts with implementations that don't account for non ascii characters and other edge cases. Whether an `AK::StringUtils::starts_with()` accounts for all edge cases they want to avoid here I don't know.

I'd say this is a case of sacrificing potential performance optimizations for correctness and spec compliance for now